clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp

```diff
@@ -8,50 +8,164 @@
 #ifndef __DPCT_CODEPIN_HPP__
 #define __DPCT_CODEPIN_HPP__
 
-#include "detail/dispatcher.hpp"
+#include "serialization/serial_basic.hpp"
+#include <fstream>
+#include <iomanip>
+#include <map>
 #include <memory>
+#include <numeric>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+namespace dpct {
+namespace experimental {
+
+namespace detail {
+
+inline static std::map<std::string, int> api_index;
+inline static std::string dump_file = "dump_log.json";
+
+class Logger {
+public:
+  Logger(const std::string &dump_file) : dst_output(dump_file) {
+    opf.open(dst_output);
+    ss << "[";
+  }
+
+  ~Logger() {
+    ss << "]";
+    opf << ss.str();
+    opf.close();
+  }
+
+  std::stringstream &get_outputstream() { return this->ss; }
+
+private:
+  std::string dst_output;
+  std::ofstream opf;
+  std::stringstream ss;
+};
+
+inline static Logger log(dump_file);
+
+inline std::map<void *, uint32_t> &get_ptr_size_map() {
+  static std::map<void *, uint32_t> ptr_size_map;
+  return ptr_size_map;
+}
+
+inline uint32_t get_pointer_size_in_bytes_from_map(void *ptr) {
+  const std::map<void *, uint32_t> &ptr_size_map = get_ptr_size_map();
+  const auto &it = ptr_size_map.find(ptr);
+  return (it != ptr_size_map.end()) ? it->second : 0;
+}
+
+inline bool is_dev_ptr(void *p) {
 #ifdef __NVCC__
-#include <cuda_runtime.h>
+  cudaPointerAttributes attr;
+  cudaPointerGetAttributes(&attr, p);
+  if (attr.type == cudaMemoryTypeDevice)
+    return true;
+  return false;
 #else
-#include <dpct/dpct.hpp>
-#include <sycl/sycl.hpp>
+  dpct::pointer_attributes attributes;
+  attributes.init(p);
+  if (attributes.get_device_pointer() != nullptr)
+    return true;
+  return false;
 #endif
-namespace dpct {
-namespace experimental {
+}
 
+template <class T>
+class TT<T, typename std::enable_if<std::is_pointer<T>::value>::type> {
+public:
+  static void dump(std::ostream &ss, T value,
+                   dpct::experimental::StreamType stream) {
+    int size = get_pointer_size_in_bytes_from_map(value);
+    size = size == 0 ? 1 : size / sizeof(*value);
```

Comment:
The size in the get_pointer_size_in_bytes_from_map return and the count of the T is different in this context, I prefer to use a new variable here.